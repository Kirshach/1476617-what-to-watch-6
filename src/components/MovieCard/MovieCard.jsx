import React, {useState, useEffect} from "react";
import {history} from '../../history';

import VideoPlayer from '../VideoPlayer/VideoPlayer';

import {filmPropTypes} from '../Film/_prop-types';
import {getFilmUrl} from '../../utils';

const TIMEOUT_BEFORE_PLAYING_PREVIEW = 500;

const MovieCard = ({film}) => {
  const [isHovered, setIsHovered] = useState(false);
  const [isPlaying, setIsPlaying] = useState(false);

  useEffect(() => {
    let clearingFunction;

    if (isHovered) {
      const playerTimeout = setTimeout(() => {
        setIsPlaying(true);
      }, TIMEOUT_BEFORE_PLAYING_PREVIEW);
      clearingFunction = () => {
        clearTimeout(playerTimeout);
      };
    }
    return clearingFunction;
  }, [isHovered]);

  const handleMouseOver = () => {
    setIsHovered(true);
  };

  const handleMouseOut = () => {
    setIsHovered(false);
    setIsPlaying(false);
  };

  return (
    <article
      className="small-movie-card catalog__movies-card"
      onMouseOver={handleMouseOver}
      onMouseOut={handleMouseOut}
      onClick={() => history.push(getFilmUrl(film.id))}
      style={{cursor: `pointer`}}
    >
      <div className="small-movie-card__image">
        <VideoPlayer
          alt={film.name}
          height={175}
          isPlaying={isPlaying}
          poster={film.previewImage}
          src={film.videoLink}
          width={280}
        />
      </div>
      <h3 className="small-movie-card__title"
        style={{pointerEvents: `none`}}
      >
        <span className="small-movie-card__link">
          {film.name}
        </span>
      </h3>
    </article>
  );
};

MovieCard.propTypes = {
  film: filmPropTypes
};

export default React.memo(
    MovieCard,
    ({film: prevFilm}, {film}) => prevFilm.id === film.id
);
