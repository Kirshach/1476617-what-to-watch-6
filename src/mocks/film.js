export const macbeth = {
  name: `Macbeth`,
  posterImage: `https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/Macbeth.jpg`,
  previewImage: `https://assets.htmlacademy.ru/intensives/javascript-3/film/preview/macbeth.jpg`,
  backgroundImage: `https://assets.htmlacademy.ru/intensives/javascript-3/film/background/Macbeth.jpg`,
  backgroundColor: `#F1E9CE`,
  description: `Macbeth, the Thane of Glamis, receives a prophecy from a trio of witches that one day he will become King of Scotland. Consumed by ambition and spurred to action by his wife, Macbeth murders his king and takes the throne for himself.`,
  rating: 3.3,
  scoresCount: 48798,
  director: `Justin Kurzel`,
  starring: [
    `Michael Fassbender`,
    `Marion Cotillard`,
    `Jack Madigan`
  ],
  runTime: 113,
  genre: `Drama`,
  released: 2015,
  id: 1,
  isFavorite: true,
  videoLink: `http://peach.themazzone.com/durian/movies/sintel-1024-surround.mp4`,
  previewVideoLink: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`
};

export const bronson = {
  name: `Bronson`,
  posterImage: `https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/bronson.jpg`,
  previewImage: `https://assets.htmlacademy.ru/intensives/javascript-3/film/preview/bronson.jpg`,
  backgroundImage: `https://assets.htmlacademy.ru/intensives/javascript-3/film/background/bronson.jpg`,
  backgroundColor: `#73B39A`,
  description: `A young man who was sentenced to seven years in prison for robbing a post office ends up spending three decades in solitary confinement. During this time, his own personality is supplanted by his alter-ego, Charles Bronson.`,
  rating: 3.6,
  scoresCount: 109661,
  director: `Nicolas Winding Refn`,
  starring: [
    `Tom Hardy`,
    `Kelly Adams`,
    `Luing Andrews`
  ],
  runTime: 92,
  genre: `Action`,
  released: 2008,
  id: 9,
  isFavorite: false,
  videoLink: `http://peach.themazzone.com/durian/movies/sintel-1024-surround.mp4`,
  previewVideoLink: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`
};

export const revenant = {
  name: `The Revenant`,
  posterImage: `https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/Revenant.jpg`,
  previewImage: `https://assets.htmlacademy.ru/intensives/javascript-3/film/preview/revenant.jpg`,
  backgroundImage: `https://assets.htmlacademy.ru/intensives/javascript-3/film/background/Revenant.jpg`,
  backgroundColor: `#92918B`,
  description: `A frontiersman on a fur trading expedition in the 1820s fights for survival after being mauled by a bear and left for dead by members of his own hunting team.`,
  rating: 4,
  scoresCount: 618498,
  director: `Alejandro G. Iñárritu`,
  starring: [
    `Leonardo DiCaprio`,
    `Tom Hardy`,
    `Will Poulter`
  ],
  runTime: 156,
  genre: `Action`,
  released: 2015,
  id: 8,
  isFavorite: false,
  videoLink: `http://peach.themazzone.com/durian/movies/sintel-1024-surround.mp4`,
  previewVideoLink: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`
};
