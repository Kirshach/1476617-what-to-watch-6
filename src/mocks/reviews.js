export const reviewData = {
  id: 1,
  user: {
    id: 13,
    name: `Zak`
  },
  rating: 1.4,
  comment: `This movie is just plain bad. There must be some big payola going round this awards season. Badly written, average acting at best, all the characters are unrelatable and inlikeable. 2 hours of my life wasted.`,
  date: `2021-03-07T08:04:28.658Z`
};

const reviewsData = [
  reviewData,
  {
    id: 2,
    user: {
      id: 17,
      name: `Emely`
    },
    rating: 7.2,
    comment: `This movie is just plain bad. There must be some big payola going round this awards season. Badly written, average acting at best, all the characters are unrelatable and inlikeable. 2 hours of my life wasted.`,
    date: `2021-02-22T08:04:28.658Z`
  },
];

export const reviews = {
  id: 1,
  data: reviewsData,
};
