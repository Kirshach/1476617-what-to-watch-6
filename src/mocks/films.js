import {macbeth, bronson, revenant} from './film';

export const films = [
  bronson,
  revenant,
  macbeth,
];
