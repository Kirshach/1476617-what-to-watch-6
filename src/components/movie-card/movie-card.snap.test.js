import React from 'react';
import {render} from '@testing-library/react';

import {macbeth} from '../../mocks/film';

import MovieCard from './movie-card';

function MockVideoPlayer() {
  return <div>This is a VideoPlayer placeholder</div>;
}

jest.mock(`../video-player/video-player`, () => ({
  __esModule: true,
  default: MockVideoPlayer,
}));

test(`<MovieCard /> renders correctly`, () => {
  const {container} = render(<MovieCard film={macbeth} />);
  expect(container).toMatchSnapshot();
});
