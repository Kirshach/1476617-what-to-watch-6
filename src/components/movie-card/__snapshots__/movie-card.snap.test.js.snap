// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MovieCard /> renders correctly 1`] = `
<div>
  <article
    class="small-movie-card catalog__movies-card"
    style="cursor: pointer;"
  >
    <div
      class="small-movie-card__image"
    >
      <div>
        This is a VideoPlayer placeholder
      </div>
    </div>
    <h3
      class="small-movie-card__title"
      style="pointer-events: none;"
    >
      <span
        class="small-movie-card__link"
      >
        Macbeth
      </span>
    </h3>
  </article>
</div>
`;
